src/storage/txn/scheduler.rs

```diff
@@ -1238,6 +1263,41 @@ impl<E: Engine, L: LockManager> Scheduler<E, L> {
     }
 }
 
+async fn get_apiv2_ctx(
+    causal_ts_provider: Option<Arc<CausalTsProviderImpl>>,
+    concurrency_manager: ConcurrencyManager,
+    max_ts_synced: bool,
+    cmd: &Command,
+) -> Result<(Option<TimeStamp>, Option<KeyHandleGuard>), Error> {
+    if causal_ts_provider.is_some() {
+        match cmd {
+            Command::RawCompareAndSwap(_) | Command::RawAtomicStore(_) => {
+                if !max_ts_synced {
+                    return Err(ErrorInner::MaxTimestampNotSynced {
+                        region_id: cmd.ctx().get_region_id(),
+                        start_ts: TimeStamp::zero(),
+                    }
+                    .into());
+                }
+                let lock_guard = get_raw_key_guard(&causal_ts_provider, concurrency_manager)
+                    .await
+                    .map_err(|err: StorageError| {
+                        ErrorInner::Other(box_err!("failed to key guard: {:?}", err))
+                    })?;
+                let ts =
+                    get_causal_ts(&causal_ts_provider)
+                        .await
+                        .map_err(|err: StorageError| {
+                            ErrorInner::Other(box_err!("failed to get casual ts: {:?}", err))
+                        })?;
+                return Ok((ts, lock_guard));
```

Comment:
`ts` & `lock_guard` must be `Some()` here.
Suggest to try using just one tier of `Option`. Two tier is a little too complicated.
`raw_ext: Option<RawExt {TimeStamp, KeyHandleGuard}>` or `raw_ext: RawExt {Option<TimeStamp>, Option<KeyHandleGuard>}`

Reply:
fixed, thanks.